### PR TITLE
docs: update issue template and Gotchas guide

### PR DESCRIPTION
**Pull request by CaerusKaru**
The Gotchas guide hasn't been updated in quite some time. This
refresh adds a more structured layout with explicit examples
and solutions for the most common issues.

This also updates the issue template to add a note about what
constitutes an appropriate issue, and a link to the gotchas guide.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/Bug.md [hunk 1]
```diff
@@ -1,3 +1,15 @@
+<--
+**NOTE**: If you provide an incomplete template, the Universal team is liable to close your issue sight on scene.
```
Comment (alan-agius4) [inline .github/ISSUE_TEMPLATE/Bug.md:2]: I find the wording here to be a bit too on the aggressive side.

**NOTE**:  Please follow the below issue template. Issues not following the issue template will be closed.

File: .github/ISSUE_TEMPLATE/Bug.md [hunk 2]
```diff
@@ -1,3 +1,15 @@
...
+Please provide _as much detail as possible_ in as minimal a reproduction as possible. This means no third-party
```
Comment (alan-agius4) [inline .github/ISSUE_TEMPLATE/Bug.md:3]: I think this should be integrated in the reproduction section."

Also not sure about no libraries etc.. there are many valid reasons why a library should be added in a reproduction.

File: .github/ISSUE_TEMPLATE/Bug.md [hunk 3]
```diff
@@ -12,11 +24,10 @@ about: Something is broken in Angular Universal
 - [ ] common
 - [ ] express-engine
 - [ ] hapi-engine
-- [ ] module-map-ngfactory-loader
 
 ### Is this a regression?
 
-<!-- Did this behavior use to work in the previous version? -->
+<!-- Did this behavior used to work in the previous version? -->
```
Comment (alan-agius4) [inline .github/ISSUE_TEMPLATE/Bug.md:30]: Nice catch

File: docs/gotchas.md [hunk 4]
```diff
@@ -1,37 +1,231 @@
...
+Although the goal of the Universal project is the ability to seamlessly render an Angular
+application on the server, there are some inconsistencies that you should consider. First,
+there is the obvious discrepancy between the server and browser environments. When rendering
+on the server, your application is in an ephemeral or "snapshot" state. The application is
+fully rendered once, with the resulting HTML returned, and the remaining application state
+destroyed until the next render. Next, the server environment inherently does not have the
+same capabilities as the browser (and has some that likewise the browser does not). For
+instance, the server does not have any concept of cookies. You can polyfill this and other
+functionality, but there is no perfect solution for this. In later sections, we'll walk
+through potential mitigations to reduce the error plane when rendering on the server.
```
Comment (alan-agius4) [inline docs/gotchas.md:14]: error plane?

File: docs/gotchas.md [hunk 5]
```diff
@@ -1,37 +1,231 @@
...
+to accomplish this in later sections.
+
+## "window is not defined"
+
+One of the most common issues when using Angular Universal is the lack of browser global
+variables in the server environment. This is because the Universal project uses
+[domino](https://github.com/fgnass/domino) as the server DOM rendering engine. As a result,
+there is certain functionality that won't be present or supported on the server. This
+includes the `window` and `document` global objects, cookies, certain HTML elements (like canvas),
+and several others. There is no exhaustive list, so please be cognizant of the fact that if you
```
Comment (alan-agius4) [inline docs/gotchas.md:28]: cognizant -> aware. To make it easier for non fluent english speakers.

File: docs/gotchas.md [hunk 6]
```diff
@@ -1,37 +1,231 @@
...
+
+  createElement(tag: string): HTMLElement {
+    return this._doc.createElement(tag);
+  }
+}
+```
+
+Please be judicious about using these references, and lower your expectations about their capabilities. `localStorage`
+is one frequently-requested API that won't work how you want it to out of the box. If you need to write your own library 
+components, please consider using this method to provide similar functionality on the server (this is what Angular CDK
```
Comment (alan-agius4) [inline docs/gotchas.md:68]: Maybe a link to the Material code?

File: docs/gotchas.md [hunk 7]
```diff
@@ -1,37 +1,231 @@
...
+
+> A Zone.js macrotask is just a JavaScript macrotask that executes in/is patched by Zone.js
+
+This means that if there is a process, like a microtask, that takes up ticks to complete, or a long-standing
+HTTP request, the rendering process will not complete, or will take longer. Macrotasks include calls to globals
+like `setTimeout` and `setInterval`, and `Observables`. Calling these without cancelling them, or letting them run 
+longer than needed on the server could result in suboptimal rendering.
+
+> It may be worth brushing up on the JavaScript event loop and learning the difference between microtasks
+> and macrotasks, if you don't know it already. [Here's](https://javascript.info/event-loop) a good reference.
```
Comment (alan-agius4) [inline docs/gotchas.md:222]: 👍

File: docs/gotchas.md [hunk 8]
```diff
@@ -1,37 +1,231 @@
...
+#### Strategy 2: Guards
+
+If you can't inject the proper global value you need from the Angular platform, you can "guard" against
+invocation of browser code, so long as you don't need to access that code on the server. For instance,
+often invocations of the global `window` element are to get window size, or some other visual aspect.
+However, on the server, there is no concept of "screen", and so this functionality is rarely needed.
+
+You may read online and elsewhere that the recommended approach is to use `isPlatformBrowser` or
+`isPlatformServer`. This guidance is **incorrect**. Instead, you should use Angular's Dependency Injection (DI)
+in order to remove the offending code and drop in a replacement at runtime. Here's an example:
```
Comment (jdpnielsen) [inline docs/gotchas.md:80]: An explanation of why this is best practice, would be great. Even just a link to some details would be nice.

File: docs/gotchas.md (hunk 6 above)
Comment (CaerusKaru) [reply docs/gotchas.md:68]: I'm hesitant to link to code directly in docs. I think it's enough to just say others do this, and using the above reference people should be able to work it out.

**Pull request merged**